### ISSUE TITLE
TransactionPublisher does not commit transaction when wrapped in reactor's Mono

### ISSUE THREAD

**Issue by juggernaut0** [jOOQ/jOOQ#13802]
### Expected behavior

After a transactionPublisher has been executed (e.g. blocked on) and completes normally, the updates should be committed and visible to later queries.

### Actual behavior

transactionPublisher does not seem to commit, meaning updates are lost.

It also logs all sorts of `io.r2dbc.spi.R2dbcNonTransientResourceException: Connection is broken: "session closed"` errors but these don't seem to bubble up as exceptions in the test.

### Steps to reproduce the problem

[MCVE](https://github.com/juggernaut0/jOOQ-mcve/blob/main/jOOQ-mcve-kotlin/src/test/kotlin/org/jooq/mcve/test/kotlin/KotlinTest.kt)

I only edited the Kotlin project because I doubt this is related to the choice of language.

### Versions

- jOOQ: 3.17.2
- Java: openjdk17
- Database (include vendor): have tested against h2 and postgres 12
- OS: ubuntu 20.04
- JDBC Driver (include name if unofficial driver): r2dbc-h2 0.9.1.RELEASE, r2dbc-postgres 0.9.1.RELEASE

**Comment by ericghara** [comment]
I'd also like to report that I'm experiencing this issue.  I'm seeing it with insert but not delete so it may not be a duplicate of #138255.

Versions
    jOOQ: 3.17.2
    Java: OpenJdk 17
    Database (include vendor): postgres 14
    OS: Gentoo Linux
    JDBC Driver (include name if unofficial driver): r2dbc-postgresql:1.0.0.RC1

**Comment by lazystone** [comment]
I'm not sure if my case is relevant or not, but anyway...

I'm building some PoC in quarkus/jooq/r2bc and was also having a problem that `transactionPublisher` worked in one case(on `select`), but not in the other(`delete/insert`). I almost gave up but then decided to convert my `Multi`(Mutiny analogue of `Flux`) to hot stream and that fixed transaction commit:

```java
  public Uni<UUID> createSession(SessionKey sessionKey) {
    var sessionId = UUID.randomUUID();
    return Multi.createFrom()
        .safePublisher(
            dsl.transactionPublisher(
                cfg -> {
                  var ctx = cfg.dsl();
                  return ctx.batch(
                      setCustomerIdQuery(ctx, sessionKey.customerId()),
                      deleteByUserQuery(ctx, sessionKey.userId()),
                      insertTranscriptionQuery(ctx, sessionKey, sessionId));
                }))
         // Following line enables transaction commit
        .toHotStream()
        .onFailure()
        .invoke(e -> logger.error("I failed!!!", e))
        .toUni()
        .replaceWith(sessionId);
  }
```

In Reactor analogue of this would be calling `.publish()` on a `Flux`.

So my theory is that since it starts working on a hot stream, then it's probably something wrong with how subscription is handled?...

I hate "reactive" 😭

**Comment by ericghara** [comment]
>  I almost gave up 
I was in the same place! If you decide not to use Jooq's transaction management,  I made a demo project to using [Spring's transaction management](https://github.com/ericghara/reactive-transactions), that _may_ be relevant to Quarkus (Idk).  Some good work-arounds were also discussed in this [GitHub issue](https://github.com/jOOQ/jOOQ/issues/12218).

**Comment by lazystone** [comment]
I decided to ditch r2bc(too unreliable) and go with good old blocking code - sooner or later project Loom will go GA. I have big hopes on it.

**Comment by lukaseder** [comment]
> I hate "reactive" 😭
> [...]
> I decided to ditch r2bc(too unreliable) and go with good old blocking code - sooner or later project Loom will go GA. I have big hopes on it.

If reactive isn't a "must have" (and it hardly ever is), then I really recommend folks don't use it!

I've never really understood the hype. Apart from bugs like these (which should be fixed, obviously!), it seems like such a footgun, for no obvious gains.

**Comment by lazystone** [comment]
off-topic obviously :) but

@lukaseder I see reactive as "inevitable evil" for **now**. It does solve some problems, but brings yet another level of complexity. And if I understand things correctly, then project Loom promise to solve most of those problems in a different and much better way. So, fingers crossed...

And, btw, thanks for JOOQ!

**Comment by lukaseder** [comment]
> I see reactive as "inevitable evil" for **now**.

Why?

**Comment by lazystone** [comment]
Ok, when *you* ask, then I start to think that it's not so "inevitable" 😄 . But "reactive" offers some way to build let's say API gateway which can handle more requests on the same resources than blocking code(and yes, I know that it's possible to use thread pools and etc.). Plus it offers some API to combine async results(but after writing some `Mono.zip` I might say it's not that better than combining `Future`s).

I did spend some time investigating thread pools starvation in some legacy apps, and then you see an article saying: "Hey! You see this new and shiny thing? It's reactive, it'll solve all you problems!". So, did I have a choice? Of course I didn't 😅 - I had to try it.

So, maybe "reactive" in java world is yet another hype as Hibernate for example(wink-wink) - it's definitely possible to write decent things on Hibernate and on "reactive"(Netty), but the cost though...

**Comment by lukaseder** [comment]
> But "reactive" offers some way to build let's say API gateway which can handle more requests on the same resources than blocking code

Did you benchmark this, and can you confirm that you got the expected improvements over blocking models, despite having I/O heavy operations, like database requests? I'm curious to see some evidence. I can imagine that in some cases, things can be implemented in the reactive way that would be extremely hard otherwise. But having played my share of [factorio](https://www.factorio.com/) (the ultimate reactive platform), I feel like there are way more "accidents," where the reactive model creates a big mess that would not happen in the blocking world, so things might actually be *worse* on average, especially when deadlines are involved.

I do agree it's a nice challenge, though.

> So, maybe "reactive" in java world is yet another hype as Hibernate for example(wink-wink) - it's definitely possible to write decent things on Hibernate and on "reactive"(Netty), but the cost though...

Well, in my opinion, Hibernate had to check the reactive box just like jOOQ did, because customers requested it, and no one will say no to customers. jOOQ won't educate end users, it will never judge, but I might :)

In the case of Hibernate, in my opinion, reactive doesn't play too well with the idea of fetching an object graph atomically, before persisting it again. Fetching parts of it, and then later on, fetching other parts produces the dreaded N+1 issue, probably much harder in the reactive world. E.g. if you fetch a parent entity, and the collection holding child entities isn't completely cached yet, then why do the whole Hibernate 1st/2nd level cache thing at all? Unless, all your child collections are also modelled as reactive streams, e.g.

```java
@Entity
class Author {
  @ManyToMany
  Flux<Book> books;
}
```

But that is really hairy, and I don't think it's supported by Hibernate?

With stateless sessions, it makes sense, of course, then it works just like jOOQ / R2DBC native. But people will probably still want *both* from Hibernate. Reactive entity graph persistence.

Anyway. This is all a digression from the bug itself, which needs to be fixed irrespective of architecture preferences :) I just took the opportunity to ask about evidence, which I'm still very curious to see.

**Comment by lazystone** [comment]
@lukaseder evidence? Are you kidding me?! Of course I don't have any 😓

Talking about evidences, I think it's possible to find some benchmarks for reactive and blocking Rest APIs. Usually they show that reactive can handle much more simultaneous requests. But then, again, you're right that if DB is involved, then things become more complicated(especially if we remember that you probably don't want to have thousands connections to DB, but rather use DB connection pool). So, I should admit, that I personally haven't done any benchmarks of my services, I just followed the hype.

It takes time to understand some things and what is **the actual** problem - blocking code isn't a problem by itself. Nothing wrong with it. The problem is with how low level code consumes resources(OS threads). So, looking back, I think that "reactive" approach tried to solve low-level problem by high-level tools, so, basically it was a workaround. Project Loom in turn aim to solve **the** problem. So it's a solution.

**Comment by lukaseder** [comment]
Finally looking into this now. The `Connection is broken: "session closed"` error is caused by the `COMMIT` call itself, which is run on a "session that has already been closed," which is why the transaction wasn't committed.

**Comment by lukaseder** [comment]
Debugging through this, the `MonoNext$NextSubscriber::onNext` call seems to cancel the `R2DBC$TransactionSubscription` first, before the transaction itself gets completed. That `cancel()` call triggers the `Connection::close` call.

So, this is a matter of correctly sequencing the events, which didn't appear in the expected order.

**Comment by venmani** [comment]
I tried this with 3.17.1 version and its working as expected. 

Mono.from(dsl.transactionPublisher(configuration -> { <code for multiple update query with transactional>}))

As of now (till going to production), I downgrade jooq version to 3.17.1 and working on, Though I face below issues. 
#13739 | Reactive transactions don't call Connection::close
#13740 | Connection::close call is skipped when R2DBC query fails

**Comment by lukaseder** [comment]
Finally, I'm confident of a fix here. Indeed, when using `Mono`, we get a `org.jooq.impl.R2DBC.AbstractSubscription.cancel()` call with this stack:

```
R2DBC$QuerySubscription<T,Q>(R2DBC$AbstractSubscription<T>).cancel() line: 191	
MonoNext$NextSubscriber<T>.onNext(T) line: 81	
0x0000000800736d68.accept(Object) line: not available	
Internal$1.onNext(T) line: 455	
R2DBC$Forwarding<T>.onNext(T) line: 237	
R2DBC$RowCountSubscriber.lambda$1(R2DBC$Forwarding, Object) line: 317	
```

The `MonoNext$NextSubscriber<T>.onNext(T)` body looks like this:

```java
@Override
public void onNext(T t) {
	if (done) {
		Operators.onNextDropped(t, actual.currentContext());
		return;
	}

	s.cancel();
	actual.onNext(t);
	onComplete();
}
```

The sequencing was a surprise to me, I would have thought delegating `onNext(t)` and `onComplete()` would precede the cancellation, but I guess there are some optimisation use-cases to make sure the upstream stops producing things as early as possible?

Anyway, it's possible to avoid the `Connection::close` call whenever we're in a `TransactionSubscription` and haven't issued `commit` or `rollback` yet. Changing this might subtly change behaviour for some users. In a stream that is not a `Mono` (e.g. a `Flux`), if the stream would produce 10 inserts, but we only `request(5)`, then a `cancel()`, those 5 will be committed, or the 10, depending on if some buffering in the stream eager executes the remaining inserts.

I guess that's what users want / expect?

**Comment by lukaseder** [comment]
The fix seems to work in all supported RDBMS except MySQL, where the test is now flaky:

- Works if run alone
- Doesn't work if run with other tests

Will look into this last bit, but this is definitely promising.

**Comment by juggernaut0** [comment]
Thanks for your work on this!

> if the stream would produce 10 inserts, but we only request(5), then a cancel(), those 5 will be committed

This makes sense to me, although buffering always makes stuff weird. 

> the test is now flaky

Oh no! Curse you MySQL! 😛

**Comment by lukaseder** [comment]
> This makes sense to me, although buffering always makes stuff weird.

All of this reactive stuff is a footgun, so why not add one more?

> Oh no! Curse you MySQL! 😛

I doubt it's MySQL's fault. It's more likely related to the no longer maintained, unofficial driver for MySQL, which is still on an 8.x version of the SPI

**Comment by lukaseder** [comment]
Well, might as well be some test artifact, where the commit is somehow flushed too slowly. The assertion of the transaction having worked correctly is run on a JDBC connection, whereas the transaction is run on an R2DBC transaction.

When running the test standalone, the JVM isn't hot yet, some classes might still be loaded, which might produce sufficient delay.

**Comment by lukaseder** [comment]
Yet again, this did "fix" it:

```java
try {
    assertAuthorIds(1, 2, 3, 4);
}
catch (AssertionError e) {
    sleep(200);
    assertAuthorIds(1, 2, 3, 4);
}
```

**Comment by lukaseder** [comment]
I will re-iterate this last MySQL issue at some later time, in case it pops up in user code. The delay really hints at something going on in the driver / protocol.

Fixed in jOOQ 3.18.0 and 3.17.7 (https://github.com/jOOQ/jOOQ/issues/14497).

Will check early on next week if https://github.com/jOOQ/jOOQ/issues/13825 is really a duplicate of this, and if it is, will soon cut a release. People have been waiting for a fix long enough.

**Comment by lukaseder** [comment]
3.17.7 has been released. Your feedback would be greatly appreciated.
